### PR TITLE
helpful credentials error message

### PR DESCRIPTION
**Pull request by danielnorberg**
Display a helpful error message telling the user how to set up credentials when getting a `401 Unauthorized` from Styx and there are no credentials available:

```
API error: Unauthorized: Please set up Application Default Credentials or set the GOOGLE_APPLICATION_CREDENTIALS env var to point to a file defining the credentials.

Hint: Try setting up credentials using gcloud: 

	$ gcloud auth application-default login
```

### CODE REVIEW

**Comment by danielnorberg** [comment]
@ulzha PTAL

File: styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java [hunk 1]
```diff
@@ -375,25 +389,29 @@
...
-      // Credential probably configured wrongly.
+    } catch (IOException | GeneralSecurityException e) {
+      // Credential probably invalid, configured wrongly or the token request failed.
       throw new RuntimeException(e);
     }
   }
 
   private CompletionStage<Response<ByteString>> executeRequest(final Request request) {
-    return client.send(decorateRequest(request)).handle((response, e) -> {
+    final Try<GoogleCredential> credential = Try.of(credentialSupplier);
```
Comment (honnix) [inline styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java]: Feels a bit weird leaking GoogleCredential in this class. Is it possible to have getToken indicating credential not configured? Eg. throwing a more specific exception than IOException.

Comment (honnix) [reply styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java]: And when that happens we don’t decorate the request.

**Comment by honnix** [comment]
👍

File: styx-cli/src/test/java/com/spotify/styx/cli/CliMainTest.java [hunk 2]
```diff
@@ -120,6 +126,34 @@ public void testWorkflowDelete() throws Exception {
...
+    when(client.triggerWorkflowInstance(any(), any(), any()))
+        .thenReturn(exceptionallyCompletedFuture(new ApiErrorException("foo", 401, false)));
+
+    try {
+      CliMain.run(cliContext, "t", "foo", "bar", "2017-01-02");
+      fail();
+    } catch (CliExitException ignored) {
+    }
+
+    verify(cliOutput).printError(Mockito.contains("gcloud auth application-default login"));
```
Comment (honnix) [inline styx-cli/src/test/java/com/spotify/styx/cli/CliMainTest.java]: `import static org.mockito.Mockito.contains` like others?

File: styx-client/src/main/java/com/spotify/styx/client/auth/GoogleIdTokenAuth.java [hunk 3]
```diff
@@ -33,21 +33,30 @@
...
 
-  GoogleIdTokenAuth(HttpTransport httpTransport) {
+  private final HttpTransport httpTransport;
+  private final Optional<GoogleCredential> credential;
+
+  GoogleIdTokenAuth(
+      HttpTransport httpTransport,
+      Optional<GoogleCredential> credential) {
     this.httpTransport = httpTransport;
+    this.credential = credential;
```
Comment (honnix) [inline styx-client/src/main/java/com/spotify/styx/client/auth/GoogleIdTokenAuth.java]: non-null check

File: styx-client/src/main/java/com/spotify/styx/client/auth/GoogleIdTokenAuth.java [hunk 4]
```diff
@@ -33,21 +33,30 @@
...
   }
 
-  public GoogleIdTokenAuth() {
-    this(Utils.getDefaultTransport());
+  public Optional<String> getToken(String targetAudience)
+      throws IOException, GeneralSecurityException {
+    return credential.isPresent()
```
Comment (honnix) [inline styx-client/src/main/java/com/spotify/styx/client/auth/GoogleIdTokenAuth.java]: can be simplified with a `map` I think.

File: styx-client/src/test/java/com/spotify/styx/client/StyxApolloClientTest.java [hunk 5]
```diff
@@ -69,8 +78,9 @@
   private static final String CLIENT_HOST = API_URL.scheme() + "://" + API_URL.host() ;
 
   @Before
-  public void setUp() throws Exception {
+  public void setUp() throws Throwable {
```
Comment (honnix) [inline styx-client/src/test/java/com/spotify/styx/client/StyxApolloClientTest.java]: Why the change here?

File: styx-cli/src/main/java/com/spotify/styx/cli/CliMain.java [hunk 6]
```diff
@@ -274,7 +275,23 @@ private void run(CliContext cliContext) {
...
+            cliOutput.printError(
+                "API error: Unauthorized: Please set up Application Default Credentials or set the "
+                    + "GOOGLE_APPLICATION_CREDENTIALS env var to point to a file defining the "
+                    + "credentials.\n"
+                    + "\n"
+                    + "Hint: Try setting up credentials using gcloud: \n"
+                    + "\n"
+                    + "\t$ gcloud auth application-default login");
+          } else {
+            cliOutput.printError("API error: Unauthorized");
```
Comment (honnix) [inline styx-cli/src/main/java/com/spotify/styx/cli/CliMain.java]: At this moment, this should never happen, right? Just asking.

**Comment by honnix** [comment]
I like this injection much better!!!

File: styx-cli/src/main/java/com/spotify/styx/cli/CliMain.java (hunk 6 above)
Comment (danielnorberg) [reply styx-cli/src/main/java/com/spotify/styx/cli/CliMain.java]: Will happen if the service rejects the credentials. Which it will start doing for e.g. external credentials at some point. And the cli should should behave well wrt that (even "old" ones).

File: styx-cli/src/test/java/com/spotify/styx/cli/CliMainTest.java (hunk 2 above)
Comment (danielnorberg) [reply styx-cli/src/test/java/com/spotify/styx/cli/CliMainTest.java]: 👍

File: styx-client/src/main/java/com/spotify/styx/client/auth/GoogleIdTokenAuth.java (hunk 3 above)
Comment (danielnorberg) [reply styx-client/src/main/java/com/spotify/styx/client/auth/GoogleIdTokenAuth.java]: will fix for both

File: styx-client/src/main/java/com/spotify/styx/client/auth/GoogleIdTokenAuth.java (hunk 4 above)
Comment (danielnorberg) [reply styx-client/src/main/java/com/spotify/styx/client/auth/GoogleIdTokenAuth.java]: `Optional.map()` does not compose well with checked exceptions =/

File: styx-client/src/test/java/com/spotify/styx/client/StyxApolloClientTest.java (hunk 5 above)
Comment (danielnorberg) [reply styx-client/src/test/java/com/spotify/styx/client/StyxApolloClientTest.java]: Remnant from earlier change that got removed. Will revert.

File: styx-client/src/main/java/com/spotify/styx/client/auth/GoogleIdTokenAuth.java (hunk 4 above)
Comment (honnix) [reply styx-client/src/main/java/com/spotify/styx/client/auth/GoogleIdTokenAuth.java]: oh, missed that again. sorry. have being programming too much Kotlin these days.

File: styx-cli/src/main/java/com/spotify/styx/cli/CliMain.java (hunk 6 above)
Comment (honnix) [reply styx-cli/src/main/java/com/spotify/styx/cli/CliMain.java]: cool

File: styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java [hunk 7]
```diff
@@ -375,25 +384,26 @@
...
 
-  private Request decorateRequest(final Request request) {
-    return withOptionalAuth(request.withHeader("User-Agent", STYX_CLIENT_VERSION).withTtl(TTL));
+  private Request decorateRequest(
+      final Request request, final Optional<String> authToken) {
+    return request
+        .withHeader("User-Agent", STYX_CLIENT_VERSION)
+        .withTtl(TTL)
+        .withHeaders(authToken
+            .map(t -> ImmutableMap.of("Authorization", "Bearer " + authToken))
```
Comment (bergman) [inline styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java]: nobody caught this one 😃

Comment (danielnorberg) [reply styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java]: 😫 good catch!
